cdap-api/src/main/java/co/cask/cdap/api/workflow/WorkflowSpecification.java

```diff
@@ -32,21 +34,28 @@
   private final String className;
   private final String name;
   private final String description;
-  private final Map<String, WorkflowActionSpecification> customActionMap;
-  private final List<ScheduleProgramInfo> actions;
   private final Map<String, String> properties;
 
+  private List<WorkflowNode> nodes = Lists.newArrayList();
+  private Map<String, WorkflowForkSpecification> forks = Maps.newHashMap();
+  private Map<String, ScheduleProgramInfo> actions = Maps.newHashMap();
+  private Map<String, WorkflowActionSpecification> customActionMap = Maps.newHashMap();
+
   public WorkflowSpecification(String className, String name, String description,
-                                      Map<String, String> properties,
-                                      List<ScheduleProgramInfo> actions,
+                                      Map<String, String> properties, List<WorkflowNode> nodes,
+                                      Map<String, WorkflowForkSpecification> forks,
+                                      Map<String, ScheduleProgramInfo> actions,
                                       Map<String, WorkflowActionSpecification> customActionMap) {
     this.className = className;
     this.name = name;
     this.description = description;
     this.properties = properties == null ? Collections.<String, String>emptyMap() :
                                            Collections.unmodifiableMap(new HashMap<String, String>(properties));
-    this.actions = Collections.unmodifiableList(new ArrayList<ScheduleProgramInfo>(actions));
```

Comment:
So, if I understand correctly, the workflow is represented as list of nodes. Each node has an id and a type. And for each node type there is a map from id to what the actual node is. Correct? 

Is that natural? The JSON for this workflow spec would be hard to read... 
I could also see this:

- workflow is a list of nodes
- node is an abstract class which provides the node id
- different subclasses of node implement he different types. For example, a fork node has getBranches() which returns a list of list of nodes. 

Wouldn't that be more natural? It keeps everything in one nested structure that reflects the nested structure of the workflow. Or is there a good reason for doing it the way that you did it?